cvmfs/download.cc

```diff
@@ -1913,63 +1940,96 @@ bool DownloadManager::SortWrtGeoReply(
     return false;
 
-  vector<string> tmp(*input_hosts);
-  for (unsigned i = 0; i < reply_vals.size(); ++i)
-    (*input_hosts)[i] = tmp[reply_vals[i]-1];
+  for (unsigned i = 0; i < expected_size; ++i) {
+    (*reply_vals)[i] = tmp_vals[i] - 1;
+  }
   return true;
 }
 
 
 /**
- * Parses a list of ';'- and '|'-separated proxy servers for the proxy groups.
- * The empty string removes the proxy chain.
+ * Parses a list of ';'- and '|'-separated proxy servers for the proxy groups,
+ *   and a ';'-separated list of fallback proxies.
+ * The empty string for both removes the proxy chain.
  */
-void DownloadManager::SetProxyChain(const string &proxy_list) {
+void DownloadManager::SetProxyChain(const string &proxy_list,
+                                    const string &fallback_proxy_list) {
   pthread_mutex_lock(lock_options_);
 
   opt_timestamp_backup_proxies_ = 0;
   opt_timestamp_failover_proxies_ = 0;
   delete opt_proxy_groups_;
-  if (proxy_list == "") {
+  if ((proxy_list == "") && (fallback_proxy_list == "")) {
     opt_proxy_groups_ = NULL;
     opt_proxy_groups_current_ = 0;
     opt_proxy_groups_current_burned_ = 0;
+    opt_proxy_groups_fallback_ = 0;
     opt_num_proxies_ = 0;
     pthread_mutex_unlock(lock_options_);
     return;
   }
 
-  // Resolve server names in provided urls, skip DIRECT
-  vector<string> urls;  // All encountered URLs
-  vector<string> proxy_groups = SplitString(proxy_list, ';');
+  // Resolve server names in provided urls
+  vector<string> hostnames;  // All encountered hostnames
+  vector<string> proxy_groups;
+  if (proxy_list != "")
+    proxy_groups = SplitString(proxy_list, ';');
+  opt_proxy_groups_fallback_= proxy_groups.size();
   for (unsigned i = 0; i < proxy_groups.size(); ++i) {
     vector<string> this_group = SplitString(proxy_groups[i], '|');
     for (unsigned j = 0; j < this_group.size(); ++j) {
-      // Note: DIRECT strings will be "extracted" to an empty string and not
-      // resolved.
-      urls.push_back(dns::ExtractHost(this_group[j]));
+      // Note: DIRECT strings will be "extracted" to an empty string.
+      string hostname = dns::ExtractHost(this_group[j]);
+      if ((hostname == "") && (fallback_proxy_list != "")) {
+        // skip DIRECTs if there are fallback proxies
+        opt_proxy_groups_fallback_--;
+      }
+      else {
+        // Save the hostname.  Leave empty (DIRECT) names so indexes will
+        // match later.  They will not be resolved.
+        hostnames.push_back(hostname);
+      }
+    }
+  }
+  vector<string> fallback_groups;
+  if (fallback_proxy_list != "")
+    fallback_groups = SplitString(fallback_proxy_list, ';');
+  for (unsigned i = 0; i < fallback_groups.size(); ++i) {
+    string fallback_proxy = fallback_groups[i];
+    size_t barpos = fallback_proxy.find('|');
+    if (barpos != std::string::npos) {
+      // ignore anything after a vertical bar, it's not supported in fallbacks
```

Comment:
There's no need for it anymore now that round-robins are fully supported as groups.  The code has to be separate at this point anyway because of some other differences, and leaving out support here makes things simpler here and elsewhere.   In fact I would like to deprecate use of vertical bars if I could now that round-robin is supported, but because of backward compatibility it will probably never happen.  In any case I'd rather not introduce a new use for it where it isn't needed.